File: src/main/java/com/kishlaly/ta/analyze/tasks/blocks/two/Long_ScreenTwo_MACD_LastShouldBeNegative.java
package com.kishlaly.ta.analyze.tasks.blocks.two;

import com.kishlaly.ta.model.BlockResultCodeJava;
import com.kishlaly.ta.model.BlockResultJava;
import com.kishlaly.ta.model.SymbolDataJava;
import com.kishlaly.ta.model.indicators.IndicatorJava;
import com.kishlaly.ta.model.indicators.MACDJava;
import com.kishlaly.ta.utils.CollectionUtilsJava;
import com.kishlaly.ta.utils.LogJava;

import java.util.List;

import static com.kishlaly.ta.model.BlockResultCodeJava.LAST_HISTOGRAM_ABOVE_ZERO_SCREEN_2;
import static com.kishlaly.ta.model.BlockResultCodeJava.OK;

public class Long_ScreenTwo_MACD_LastShouldBeNegative implements ScreenTwoBlockJava {
    @Override
    public BlockResultJava check(SymbolDataJava screen) {
        List<MACDJava> screenTwoMacdValues = (List<MACDJava>) screen.indicators.get(IndicatorJava.MACD);
        double latestHistogramValue = CollectionUtilsJava.getFromEnd(screenTwoMacdValues, 1).getHistogram();

        if (latestHistogramValue > 0) {
            LogJava.recordCode(BlockResultCodeJava.LAST_HISTOGRAM_ABOVE_ZERO_SCREEN_2, screen);
            LogJava.addDebugLine("histogram at the right edge above zero");
            return new BlockResultJava(screen.getLastQuote(), LAST_HISTOGRAM_ABOVE_ZERO_SCREEN_2);
        }
        return new BlockResultJava(screen.getLastQuote(), OK);
    }
}

File: src/main/kotlin/com/kishlaly/ta/analyze/tasks/blocks/two/macd/Long_ScreenTwo_MACD_LastShouldBeNegative.kt
package com.kishlaly.ta.analyze.tasks.blocks.two.macd

import com.kishlaly.ta.analyze.tasks.blocks.two.ScreenTwoBlock
import com.kishlaly.ta.model.BlockResult
import com.kishlaly.ta.model.BlockResultCode
import com.kishlaly.ta.model.SymbolData
import com.kishlaly.ta.model.indicators.Indicator
import com.kishlaly.ta.model.indicators.MACD
import com.kishlaly.ta.utils.Log

class Long_ScreenTwo_MACD_LastShouldBeNegative : ScreenTwoBlock {
    override fun check(screen: SymbolData): BlockResult {
        val screenTwoMacdValues = screen.indicator(Indicator.MACD) as List<MACD>
        if (screenTwoMacdValues.last().histogram > 0) {
            Log.recordCode(BlockResultCode.LAST_HISTOGRAM_ABOVE_ZERO_SCREEN_2, screen)
            Log.addDebugLine("histogram at the right edge above zero")
            return BlockResult(screen.lastQuote, BlockResultCode.LAST_HISTOGRAM_ABOVE_ZERO_SCREEN_2)
        }
        return BlockResult(screen.lastQuote, BlockResultCode.OK)
    }
}